### PR TITLE
Reuse the last known good autoupdate settings and report fail-closed state

### LINKED ISSUE #78: Autoupdates: fall back to the last known good settings instead of disabling all updates
`AutoUpdatePluginsFilter` treats any failure to reach the OpsOasis settings route as
`disable_all => true`. That is the right *shape* for a deliberate stop, but it is also what
happens for every transient reason the route is momentarily unavailable — and the result is
that plugin, theme and core auto-updates stop across every connected site until it recovers.

This issue records the change we would like here. Nothing needs to change on the OpsOasis
side; the coupling lives in this module.

## What was measured (2026-09-01)

**A stand-down of the OpsOasis application is a fast 404, not a timeout.** Every way its
must-use plugin can decline to load — the operational kill switch, a missing
`vendor/autoload.php`, a failed requirements gate — leaves WordPress serving normally with the
route simply unregistered. So the failure arrives as a prompt `rest_no_route` 404, and the
`timeout => 2` never comes into play.

**The CDN extends the outage rather than masking it.** Probing a non-existent path under the
same namespace (`/wp-json/wpcomsp/autoupdate-plugin/v1/…`) returns MISS then HIT, HIT carrying
`cache-control: max-age=300, must-revalidate`. A 404 is a valid response rather than an error,
so stale-if-error does not apply to it: an edge node can keep serving a cached 404 for up to
five minutes *after* the origin is healthy again. Stacked on this module's own five-minute
negative transient, a site can stay in the fail-safe state for roughly **ten minutes after
the route has recovered**.

## Requested change: last known good, with a bounded grace

In `src/Modules/Autoupdates/AutoUpdatePluginsFilter.php`:

1. **Persist the last successful payload in an option, not only a transient**, so it survives
   object-cache eviction and is available exactly when the fetch is failing.
2. **On failure, fall back to that last known good payload** instead of
   `(object) array( 'disable_all' => true )`.
3. **Fall closed only after a bounded grace since the last success** (24h is a reasonable
   starting point), so a genuinely long outage still ends in the conservative state while a
   brief one costs the fleet nothing.
4. **Leave `disable_all` as the explicit way to stop the fleet.** It is already a first-class
   setting, so an availability failure does not need to imitate it.

Falling back to the last known good payload is conservative in the direction that matters: a
stale `disabled_plugins` list still blocks the plugins somebody deliberately blocked. Only a
*newly added* block would be missed during the outage window — and today, that same window
blocks everything, including security updates.

## Two smaller notes

- **The transient key is shared with the predecessor.** Both this module and
  `a8cteam51/plugin-autoupdate-filter` use `wpcpmsp_auto_update_settings`, so on a site where
  both are active they populate each other's cache. If the cached value's shape changes, either
  account for that or namespace the key.
- **Sites still on the predecessor have a worse posture.** It hooks `init` unconditionally with
  no admin/cron/CLI gate, passes no `timeout`, and never negative-caches — so while the route is
  failing it re-issues the request on every uncached request, front-end included. Retiring it in
  favour of this module is worth doing independently of the change above.

### LINKED ISSUE #82: Autoupdates fail-closed state is invisible to the fleet — /status reports the module as merely enabled
### Description

When the centralized settings endpoint cannot be reached, the Autoupdates module turns off every plugin,
core and theme autoupdate on the site, and nothing outside that site can tell that it happened.

`AutoUpdatePluginsFilter::get_auto_update_settings()` fetches
`https://opsoasis.wpspecialprojects.com/wp-json/wpcomsp/autoupdate-plugin/v1/settings/` with a **2 second
timeout**. On any failure — connection error, non-2xx response, malformed JSON — it negative-caches a
fail-closed object and rethrows:

```php
} catch ( \Exception $exception ) {
    // Negative-cache the fail-safe so a slow/down OpsOasis cannot block every uncached page load.
    set_transient( $transient_key, (object) array( 'disable_all' => true ), 5 * MINUTE_IN_SECONDS );
    throw $exception;
}
```

`initialize()` catches that and sets `$this->settings = (object) array( 'disable_all' => true )`, which
`filter_maybe_disable_all_autoupdates()` then enforces on `auto_update_plugin`, `auto_update_core` and
`auto_update_theme` at `PHP_INT_MAX` priority — so it wins over everything else.

Failing closed is the right default and is not in question here. The problem is that the resulting state is
invisible and the trigger is very cheap to hit.

**The state is not reported anywhere off-site.** `Status_Controller::get_status()` builds each module's
entry as just:

```php
$modules[ $key ] = array(
    'name'    => $module->get_name(),
    'enabled' => $module->is_active(),
);
```

`messages` and `bot-protection` get extra keys added after that loop; `autoupdates` gets nothing. So a site
that has had all autoupdates disabled for hours reports the autoupdates module as `enabled: true`, which is
also true and completely misleading. There is no aggregated signal, and the only local signal is an admin
notice that requires somebody to be looking at that site's wp-admin.

**A 2 second blip can cost a whole update cycle.** `initialize()` returns early unless
`is_autoupdate_context()` is true (`is_admin() || wp_doing_cron() || WP_CLI`), so on most sites the fetch
that matters is the one during the cron run that actually evaluates updates. If that single 2 second request
fails, that run applies nothing. The 5 minute negative cache means the site recovers quickly, but the
recovery is irrelevant if the update evaluation only happens twice a day and it happened to land inside the
failed window.

**The endpoint is hardcoded.** The URL is a literal in `get_auto_update_settings()` with no constant or
filter to override it, so there is no way to point a site at a staging endpoint or to stub it in a test.

### Steps to reproduce

1. On a site with the Autoupdates module active, block outbound requests to
   `opsoasis.wpspecialprojects.com` (or point it at `127.0.0.1` in `/etc/hosts`).
2. Run `wp cron event run wp_version_check` or load any wp-admin page.
3. Request the Atlantis REST status endpoint.

### What you expected to happen

The status payload should show that the module is currently fail-closed and not applying any updates, so a
prolonged endpoint outage is visible fleet-wide without visiting individual sites.

### What actually happened

The status payload reports `"enabled": true` for the autoupdates module and nothing else. Every plugin,
core and theme autoupdate on that site is being vetoed and there is no way to see that remotely.

### Suggested scope

1. **Add the fail-closed state to the `/status` payload.** This is the part with actual value — it turns a
   fleet-wide outage from invisible into queryable. `messages` and `bot-protection` already set the
   precedent for module-specific keys.
2. **Make the endpoint overridable** via a constant or filter, which also makes the module testable.
3. **Consider a bounded stale-cache tolerance** — keep serving the last known-good settings for some window
   rather than failing closed on the first error. This is a behaviour change to the fail-safe and deserves
   its own discussion; it should not be bundled in without agreement.

### Note on the source report

The security review lists this as using `wp_remote_get`. It does not — it uses `wp_safe_remote_get`, so the
request is already protected against redirect-based SSRF. That part of the finding is stale and no change is
needed for it.

### PR DESCRIPTION
**Pull request by gin0115**
#### Changes proposed in this Pull Request

Addresses #82 and #78 together, because they are the same code path: #78 changes what happens during an outage, and #82 makes that state visible from outside the site.

**Last known good settings, with a bounded grace (#78)**

* Every successful fetch is now recorded in a new option, `a8csp_atlantis_autoupdate_last_good_settings`, holding the payload and the time it was fetched. An option rather than only a transient, so it survives object-cache eviction — which is exactly when it is needed.
* On a failed fetch, `get_auto_update_settings()` serves that payload instead of `disable_all => true`, and returns normally rather than throwing.
* It falls closed only once the last success is older than a grace window — 24 hours by default, adjustable with the new `a8csp_atlantis_autoupdate_settings_grace` filter. A prolonged outage still ends in the conservative state.
* `disable_all` is untouched and remains the explicit way to stop the fleet. An availability failure no longer has to imitate it.

Serving stale settings is conservative in the direction that matters: a stale `disabled_plugins` list still blocks whatever somebody deliberately blocked. Only a *newly added* block is missed during the window — against which the previous behaviour blocked everything, security updates included.

**Fail-closed state in the status payload (#82)**

* `modules.autoupdates` in the REST status response gains `fail_closed`, `last_success` and `seconds_since_success`.
* These are derived from stored data via a new `AutoUpdatePluginsFilter::get_settings_state()`, not from the module's `$settings` property. That matters: a REST request is not an autoupdate context, so `initialize()` returns early and `$settings` is never set. Reading the instance would have reported nothing.
* `enabled` deliberately still reports `true` throughout an outage — that is accurate, and it is precisely why a separate signal was needed. There is a test asserting both at once.

**Endpoint override (#82)**

* The hardcoded OpsOasis URL moves behind `AutoUpdatePluginsFilter::get_settings_endpoint()`, overridable by the `A8CSP_ATLANTIS_AUTOUPDATE_SETTINGS_URL` constant or the `a8csp_atlantis_autoupdate_settings_url` filter. This is also what makes the fetch stubbable, and several of the new tests depend on it.

#### Testing instructions

```bash
npm run tests:run:integration   # 46 tests, 161 assertions
```

Seven new tests, six in `AutoupdatesTestCest` and one in a new `StatusControllerTestCest`:

| Test | Asserts |
| --- | --- |
| `settings_endpoint_is_overridable` | the request goes to the filtered URL |
| `successful_fetch_records_last_known_good_settings` | a success writes the option |
| `failed_fetch_falls_back_to_last_known_good_within_grace` | an outage keeps the stored blocks and does not set `disable_all` |
| `failed_fetch_falls_closed_once_the_grace_period_expires` | past the window, the fail-safe is negative-cached as before |
| `grace_period_is_filterable` | a zero window fails closed immediately |
| `settings_state_is_readable_without_initializing_the_module` | state is correct with nothing stored, inside the window, and past it |
| `status_payload_reports_the_autoupdates_fail_closed_state` | the REST payload carries the keys, and `enabled` stays true in both states |

Manual check:

1. Point the site at a dead endpoint: `add_filter( 'a8csp_atlantis_autoupdate_settings_url', fn() => 'https://127.0.0.1/nope' );`
2. Load wp-admin once with a good endpoint first so a last known good payload is recorded, then switch to the dead one.
3. Request `/wp-json/a8csp-atlantis/v1/status` as an administrator. `modules.autoupdates.fail_closed` is `false` while inside the grace window; autoupdates continue to be evaluated normally.
4. Set `add_filter( 'a8csp_atlantis_autoupdate_settings_grace', fn() => 0 );` and request it again. `fail_closed` is now `true`.

#### Notes

* **Behaviour change worth flagging:** during the grace window the fetch now succeeds, so the "Unable to get autoupdate settings" admin notice no longer appears for brief outages. That is the intended effect — the local signal goes quiet while the remote one becomes accurate — but it is a visible difference for anyone used to seeing that notice.
* The `wpcpmsp_auto_update_settings` transient key is deliberately left alone.
* #78's other note, about retiring the predecessor plugin on sites still running it, is out of scope here — it concerns a different plugin.
* `phpcs` is clean. This PR also adds the `@throws` tag that `phpcs` was already failing on for this file on `trunk`.
* `phpmd` reports 13 violations and `phpstan` 10 errors, all pre-existing and none in the files touched here — they are in `src/CLI/*`, `PluginFilterRules.php`, `ListTable.php`, `Bilmur.php`, `Plugin.php` and `Force_Update_Check_Controller.php`.
* The `EndToEnd` suite contains no tests, only a bootstrap, so there was nothing to run there.

Fixes #82
Fixes #78

### CODE REVIEW

File: src/REST/Status_Controller.php [hunk 1]
```diff
@@ -112,6 +113,13 @@ public function get_item( \WP_REST_Request $request ): \WP_REST_Response { // ph
 			$modules['messages']['count'] = $this->count_messages();
 		}
 
+		// `enabled` only says the module is switched on. When the centralized settings cannot be
+		// fetched the module vetoes every plugin, theme and core update while still reporting
+		// enabled, so fleet tooling needs `fail_closed` to spot a prolonged outage.
+		if ( isset( $modules['autoupdates'] ) ) {
+			$modules['autoupdates'] += AutoUpdatePluginsFilter::get_settings_state();
```
[the remark this answers is not in the document]
Comment (gin0115) [reply src/REST/Status_Controller.php]: Confirmed and fixed in 11e8e38.

Verified the chain: `Modules::try_initialize_module()` assigns `$this->modules[ $key ]` unconditionally, and `AbstractModule::maybe_initialize()` returns at the `is_active()` gate before `initialize()` runs — so a switched-off module never fetches and never writes `a8csp_atlantis_autoupdate_last_good_settings`. `a8csp_atlantis_maybe_disable_autoupdates_module_on_activation()` sets `enabled => '0'` on every site carrying legacy plugin-autoupdate-filter installed-but-inactive, so this was the common case rather than an edge case.

Fixed slightly differently from the suggestion: rather than forcing `fail_closed` to false, the keys are not added at all when the module is off. A site that vetoes nothing has no answer to give, and omitting the keys says that more clearly than a `false` that looks like a healthy measurement.

Covered by `status_payload_omits_autoupdate_state_when_the_module_is_off()`, which asserts the keys are absent while off and present while on.

File: src/Modules/Autoupdates/AutoUpdatePluginsFilter.php [hunk 2]
```diff
@@ -175,9 +196,19 @@ private function get_auto_update_settings(): \stdClass {
 
 				// Save the settings in a transient for 5 minutes
 				set_transient( $transient_key, $decoded_body, 5 * MINUTE_IN_SECONDS );
+				self::store_last_known_good( $decoded_body );
```
[the remark this answers is not in the document]
Comment (gin0115) [reply src/Modules/Autoupdates/AutoUpdatePluginsFilter.php]: Accurate as a reading of the code, but not being changed.

The scenario needs OpsOasis to answer `200` with a body that decodes to something other than an object. I have not been able to show that ever happens. The measurements recorded in #78 say the opposite: an OpsOasis stand-down is a fast `rest_no_route` 404, not an empty 200.

A 404 is already handled correctly here — the response-code check throws before `json_decode` is reached, so it lands in the `catch` and serves the last known good payload without touching stored settings. That is the path that actually occurs, and it behaves as intended.

The only evidence that a non-object body is possible at all is the pre-existing placeholder branch and its comment, which may have been written defensively rather than in response to something observed.

Leaving it. Happy to revisit if anyone can point at a real response that does this.

File: src/Modules/Autoupdates/AutoUpdatePluginsFilter.php [hunk 3]
```diff
@@ -187,6 +218,124 @@ private function get_auto_update_settings(): \stdClass {
...
+				'last_success'          => null,
+				'seconds_since_success' => null,
+			);
+		}
+
+		$timestamp = (int) $stored['timestamp'];
+		$age       = time() - $timestamp;
+
+		return array(
+			'fail_closed'           => $age >= self::get_grace_period(),
```
[the remark this answers is not in the document]
Comment (gin0115) [reply src/Modules/Autoupdates/AutoUpdatePluginsFilter.php]: Correct in principle, not being changed.

The site this describes needs no wp-admin activity, no front-end traffic and no system cron for a full 24 hours. Front-end traffic spawns WP-Cron, and `wp_doing_cron()` is true in that spawned request, so any visited site fetches. Anything through `admin-ajax.php` — heartbeat included — runs with `WP_ADMIN` defined, so any open admin tab fetches too. What is left is a dormant site.

And on a dormant site with broken cron, autoupdates genuinely are not running, so `fail_closed => true` is not a false alarm about the outcome. It is misleading about the cause: it points at OpsOasis when the problem is cron.

Recording a `last_attempt` timestamp in the `catch` block, as suggested, is the right shape for a fix and would let a failing endpoint be told apart from an idle site. Deliberately deferred — the reporting cost outweighs the case it covers.

File: tests/Integration/AutoupdatesTestCest.php [hunk 4]
```diff
@@ -343,6 +343,275 @@ public function front_end_initialize_makes_no_remote_request( IntegrationTester
...
+	/**
+	 * The fail-closed state must be readable from storage, because the REST status request
+	 * is not an autoupdate context and so never populates the module's settings property.
+	 *
+	 * @param IntegrationTester $i Tester instance.
+	 *
+	 * @return void
+	 */
+	public function settings_state_is_readable_without_initializing_the_module( IntegrationTester $i ): void {
+		$this->reset_settings_storage();
```
[the remark this answers is not in the document]
Comment (gin0115) [reply tests/Integration/AutoupdatesTestCest.php]: Correct — the cleanup is on the happy path only, and a failed assertion would leave the 25-hour-old fixture behind for whatever runs next, turning one real failure into several unrelated ones.

Not changing it in this PR by decision, not because the finding is wrong. Noting it here so it is on the record rather than lost.

**Comment by gin0115** [comment]
Hey @ahegyes

A list of issues we had actually covered a similar one you opened in #78. So i addressed them both in this PR. 

Let me know if this approach works for you?

**Comment by ahegyes** [comment]
Yeah, LGTM! Didn't expect this to be handled so quickly 😅

**Review by nickpagz — APPROVED** [review]
LGTM
